third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -543,667 +786,694 @@ enum ReadMode {
-            if e.is_mut && e.kind.overlaps(&edge.kind) {
-                if let Some(diag) = self.borrow_edge_info("previous ", state, true, e) {
-                    diags.push(diag)
+    /// Check whether a local can be written. This is only allowed if no borrowed references exist.
+    fn check_write_local(&self, local: TempIndex) {
+        if self.is_ref(local) {
+            // Always valid
+            return;
+        }
+        if let Some(label) = self.state.label_for_temp_with_children(local) {
+            // The destination is currently borrowed and cannot be assigned
+            self.error_with_hints(
+                self.cur_loc(),
+                format!("cannot assign to borrowed {}", self.display(local)),
+                "attempted to assign here",
+                self.borrow_info(label, |_| true)
+                    .into_iter()
+                    .chain(self.usage_info(label, |t| t != &local)),
+            )
+        }
+    }
+
+    /// Check whether the borrow graph is 'safe' w.r.t a set of `temps`. See the discussion of safety at the
+    /// beginning of this file.
+    ///
+    /// To effectively check the path-oriented conditions of safety here, we need to deal with the fact
+    /// that graphs have non-explicit choice nodes, for example:
+    ///
+    /// ```ignore
+    ///                 s
+    ///            &mut /\ &mut
+    ///             .f /  \ .g
+    ///              r1    r2
+    /// ```
+    ///
+    /// The diverging edges `.f` and `.g` are not directly visible. In order to deal with this, we construct a
+    /// _hyper graph_ on the fly as follows:
+    ///
+    /// 1. The root nodes are the singleton sets with all the root nodes for the given temporaries.
+    /// 2. The hyper edges are grouped into those of the same edge kind. Hence, two `&mut` edges
+    ///    like in the example above become one hyper edge. The successor state of the hyper edge
+    ///    is the union of all the targets of the edges grouped together.
+    ///
+    /// If we walk this graph now from the root to the leaves, we can determine safety by directly comparing
+    /// hyper edge siblings.
+    fn check_borrow_safety(&mut self, temps: &BTreeSet<TempIndex>) {
+        let leaves = self
+            .state
+            .leaves()
+            .into_iter()
+            .filter_map(|(l, mut ts)| {
+                ts = ts.intersection(temps).cloned().collect();
+                if !ts.is_empty() {
+                    Some((l, ts))
                 } else {
-                    debug_assert!(false, "unexpected Skip edge")
+                    None
+                }
+            })
+            .collect::<BTreeMap<_, _>>();
+        // Initialize root hyper nodes
+        let mut hyper_nodes: BTreeSet<BTreeSet<LifetimeLabel>> = BTreeSet::new();
+        for leaf in leaves.keys() {
+            for root in self.state.roots(leaf) {
+                hyper_nodes.insert(iter::once(root).collect());
+            }
```

Comment:
This graphs contain rarely more then a few nodes and have depth deeper than 3 so I think we are good. We can micro optimize if we figure via benchmarking its an issue. Moreover, the entire logic is still evolving, once the problem is fully understood (which it isn't yet), we can write v3 with fully optimal graph algorithms. This would probably also the point to create a paper.